source/common/http/http1/parser.h

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace Http {
+namespace Http1 {
+
+enum class ParserType { Legacy, LlHttp };
+
+enum class MessageType { Request, Response };
+
+/**
+ * Callbacks base class. Any users of the HTTP1 parser who want to use the parsed data needs to
+ * implement this interface and pass the object to the `data` parameter of ParserFactory::create.
+ */
+class ParserCallbacks {
+public:
+  virtual ~ParserCallbacks() = default;
+  /**
+   * Called when a request/response is beginning.
+   */
+  virtual int onMessageBegin() PURE;
+
+  /**
+   * Called when URL data is received.
+   * @param data supplies the start address.
+   * @param length supplies the length.
+   */
+  virtual int onUrl(const char* data, size_t length) PURE;
+  virtual int onStatus() PURE;
+
+  /**
+   * Called when header field data is received.
+   * @param data supplies the start address.
+   * @param length supplies the length.
+   */
+  virtual int onHeaderField(const char* data, size_t length) PURE;
+
+  /**
+   * Called when header value data is received.
+   * @param data supplies the start address.
+   * @param length supplies the length.
+   */
+  virtual int onHeaderValue(const char* data, size_t length) PURE;
+
+  /**
+   * Called when headers are complete.
+   * @return 0 if no error, 1 if there should be no body.
+   */
+  virtual int onHeadersComplete() PURE;
+
+  /**
+   * Called when body data is received.
+   * @param data supplies the start address.
+   * @param length supplies the length.
+   */
+  virtual int onBody(const char* data, size_t length) PURE;
+
+  /**
+   * Called when the request/response is complete.
+   */
+  virtual int onMessageComplete() PURE;
+  virtual int onChunkHeader() PURE;   // shrug
+  virtual int onChunkComplete() PURE; // shrug
+};
+
+/**
+ * Parser interface.
+ */
+class Parser {
+public:
+  virtual ~Parser() = default;
+  virtual int execute(const char* slice, int len) PURE;
+  virtual void resume() PURE;
+  virtual int pause() PURE;
```

Comment:
Would it make sense to return ParserStatus here? Or would it be too much work to translate from implementation specific status to the ParserStatus?